### PR TITLE
Implement Equihash in a platform-independent manner

### LINKED ISSUE #850: Specify and code the PoW algorithm in platform-independent manner.
We need to specify our PoW in a platform independent manner, and update the code to use explicit platform-independent constructs. As far as I currently understand, I think the main issue here is around the memory layout of indices as they are provided to hash inputs. Eg: https://github.com/zcash/zcash/pull/849/commits/6d25662f258d49ab8f88635d309bc1148070a46c#r59067626

### PR DESCRIPTION
**Pull request by str4d**
Closes #850.

### CODE REVIEW

File: src/crypto/equihash.cpp [hunk 1]
```diff
@@ -22,19 +22,24 @@
...
 int Equihash<N,K>::InitialiseState(eh_HashState& base_state)
 {
-    unsigned int n = N;
-    unsigned int k = K;
     unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES] = {};
-    memcpy(personalization, "ZcashPOW", 8);
-    memcpy(personalization+8,  &n, 4);
-    memcpy(personalization+12, &k, 4);
+    memcpy(personalization, "ZcashPoW", 8);
+    personalization[8]  =  N        & 0xFF;
```
Comment (ebfull) [inline src/crypto/equihash.cpp]: Let's have a function generically do this.

File: src/crypto/equihash.cpp [hunk 2]
```diff
@@ -75,7 +81,12 @@ StepRow<WIDTH>::StepRow(unsigned int n, const eh_HashState& base_state, eh_index
 {
     eh_HashState state;
     state = base_state;
-    crypto_generichash_blake2b_update(&state, (unsigned char*) &i, sizeof(eh_index));
+    unsigned char array[sizeof(eh_index)];
+    array[0] =  i        & 0xFF;
```
Comment (ebfull) [inline src/crypto/equihash.cpp]: As well as here.

**Comment by ebfull** [comment]
ACK modulo nits.

File: src/crypto/equihash.cpp (hunk 1 above)
Comment (str4d) [inline src/crypto/equihash.cpp]: Hmm... what about:

```
memcpy(personalization+8,  &(htole32(n)), 4);
memcpy(personalization+12,  &(htole32(k)), 4);
```

Comment (str4d) [inline src/crypto/equihash.cpp]: Or better:

```
uint32_t n = htole32(N);
uint32_t k = htole32(K);
mcpy(personalization+8,  &n, 4);
memcpy(personalization+12,  &k, 4);
```

This matches [how the indices get serialised in the block header](https://github.com/zcash/zcash/blob/447444ae7a3ed4e13bfb09e0c15d458070e9b904/src/serialize.h#L92).

Comment (ebfull) [inline src/crypto/equihash.cpp]:
```c++
void writeBE(unsigned char* arr, unsigned int v) {
    arr[0] = v & 0xFF;
    arr[1] = (v >> 8) & 0xFF;
    arr[2] = (v >> 16) & 0xFF;
    arr[3] = (v >> 24) & 0xFF;
}

// ...

writeBE(&personalization[8], N);
writeBE(&personalization[12], K);

// ...

writeBE(array, i);
```

Comment (str4d) [inline src/crypto/equihash.cpp]: Isn't that little-endian? I'm confused now.

Comment (ebfull) [inline src/crypto/equihash.cpp]: Whoops, rename the function.

**Comment by str4d** [comment]
Rebased on latest, and nits addressed in a new commit (previous commits unchanged). Still need to update miner tests before this can be merged.

**Comment by ebfull** [comment]
ACK

**Comment by str4d** [comment]
I have updated my [Python reference implementation](https://github.com/str4d/zcash-pow) to be consistent with this PR. The same test vectors now pass there too!

**Comment by daira** [comment]
I intend to review this today, Internet access permitting.

File: src/crypto/equihash.cpp [hunk 3]
```diff
@@ -24,10 +24,10 @@
 template<unsigned int N, unsigned int K>
 int Equihash<N,K>::InitialiseState(eh_HashState& base_state)
 {
-    unsigned int n = N;
-    unsigned int k = K;
+    uint32_t n = htole32(N);
+    uint32_t k = htole32(K);
```
Comment (daira) [inline src/crypto/equihash.cpp]: I suggest renaming these to ``le_N`` and ``le_K``.

File: src/crypto/equihash.cpp [hunk 4]
```diff
@@ -24,10 +24,10 @@
...
     unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES] = {};
-    memcpy(personalization, "ZcashPOW", 8);
+    memcpy(personalization, "ZcashPoW", 8);
```
Comment (daira) [inline src/crypto/equihash.cpp]: Well spotted.

File: src/crypto/equihash.cpp [hunk 5]
```diff
@@ -37,26 +37,21 @@ int Equihash<N,K>::InitialiseState(eh_HashState& base_state)
                                                          personalization);
 }
 
+// Big-endian so that array comparison is equivalent to integer comparison
```
Comment (daira) [inline src/crypto/equihash.cpp]: "lexicographic array comparison"

File: src/crypto/equihash.cpp [hunk 6]
```diff
@@ -37,26 +37,21 @@ int Equihash<N,K>::InitialiseState(eh_HashState& base_state)
...
     assert(sizeof(eh_index) == 4);
-    array[0] = (i >> 24) & 0xFF;
-    array[1] = (i >> 16) & 0xFF;
-    array[2] = (i >>  8) & 0xFF;
-    array[3] =  i        & 0xFF;
+    eh_index bei = htobe32(i);
+    memcpy(array, &bei, sizeof(eh_index));
 }
 
+// Big-endian so that array comparison is equivalent to integer comparison
```
Comment (daira) [inline src/crypto/equihash.cpp]: "lexicographic array comparison"

**Comment by daira** [comment]
ACK modulo nits.

**Comment by defuse** [comment]
ACK

**Comment by str4d** [comment]
Nits addressed!

@zkbot r+

**Pull request merged**